### PR TITLE
feat: NFR Tests Part 1

### PR DESCRIPTION
**Pull request by DFong298**
# GitHub Ticket / Description

[SS-149](https://github.com/lilweege/SyntaxSentinels/issues/149)

## Type && Checklist

- [ ] **New feature (new APIs, new functional methods)**
  - [ ] Created (or updated) unit test cases (if required)
  - [ ] Passed unit testing for **ALL** cases
  - [ ] Created (or updated) E2E cases (if required)
  - [ ] Passed E2E testing for **ALL** cases (local or/and sandbox)
- [ ] **Refactoring / Enhancement**
  - [ ] This is breaking change
  - [ ] Created (or updated) unit test cases (if required)
  - [ ] Passed unit testing for **ALL** cases
  - [ ] Created (or updated) E2E cases (if required)
  - [ ] Passed E2E testing for **ALL** cases (local or/and sandbox)
- [ ] **Bug-fix**
  - [ ] Type:
    - [ ] Code Issue:
    - [ ] Business Logic Issue:
  - [ ] Created (or updated) unit test cases (if required)
  - [ ] Passed unit testing for **ALL** cases
  - [ ] Created (or updated) E2E cases (if required)
  - [ ] Passed E2E testing for **ALL** cases (local or/and sandbox)

## Dependency

[Are there any required updates which cannot be done in this repo, or must be done after completing this PR?]

## Choose proper operation and replace the \<placeholder\>: `+semver:<placeholder>`

`breaking` `major` `feature` `minor` `fix` `patch`

+semver:patch

### CODE REVIEW

**Comment by lucasichen** [comment]
please fix subsections as they should be under section 4.2 Tests for Nonfunctional Requirements
image

File: docs/VnVPlan/VnVPlan.tex [hunk 1]
```diff
@@ -549,8 +549,317 @@ \subsection{Tests for Nonfunctional Requirements}
...
+These tests are to ensure that the way the front end is designed adheres to the
+defined requirements with respect to layout, typography, component alignment,
+etc. 
+
+\begin{description}
+  \item[Front End Component Review:] We will engage a developer to look through
+  the code and interface and ensure components are implemented as they are
+  defined in the SRS.
+  \item[Accessibility Check:] A developer will review if accessibility tools
+  exist or not within the application. Important tools are screen readers and
```
Comment (MohammadMohsinKhan) [inline docs/VnVPlan/VnVPlan.tex:561]: mention aria-label for screen reader

Comment (MohammadMohsinKhan) [reply docs/VnVPlan/VnVPlan.tex:561]: also do we need to specify who the developer is? just wondering how specific we should be

File: docs/VnVPlan/VnVPlan.tex [hunk 2]
```diff
@@ -549,8 +549,317 @@ \subsection{Tests for Nonfunctional Requirements}
...
+  alt text for images. Furthermore, the developer must check for sufficient
+  colour contrast between components.
+\end{description}
+
+
+\begin{enumerate}
+
+\item{NFR-UI1}
+
+NFR: LF-AR1, LF-AR2, LF-AR3, LF-SR1, LF-SR2, LF-SR3, LF-SR4
```
Comment (MohammadMohsinKhan) [inline docs/VnVPlan/VnVPlan.tex:571]: why are all of these grouped? this feels way too vague @lucasichen thoughts?

File: docs/VnVPlan/VnVPlan.tex [hunk 3]
```diff
@@ -549,8 +549,317 @@ \subsection{Tests for Nonfunctional Requirements}
...
+Type: Manual, Accessibility Check
+					
+Initial State: Application on home page, ready for use
+					
+Input: Tester engagement
+					
+Output: List of missing accessibility considerations that have been overlooked
+					
+How test will be performed: 
+\begin{enumerate}
```
Comment (MohammadMohsinKhan) [inline docs/VnVPlan/VnVPlan.tex:600]: again, same comment as above, this test is just way too vague. Mention what enables screen readers (aria-labels mainly). Mention how the "developer" should make sure alt text exists for every image

File: docs/VnVPlan/VnVPlan.tex [hunk 4]
```diff
@@ -549,8 +549,317 @@ \subsection{Tests for Nonfunctional Requirements}
...
+  Initial State: Application on home page, ready for use
+            
+  Input: Tester engagement
+            
+  Output: A count for the number of clicks required to navigate to different functionality
+            
+  How test will be performed: 
+  \begin{enumerate}
+    \item A list of commonly used functions will be created (System is not large, therefore this can be done manually)
+    \item The developer will navigate to each of those functions, and note how many clicks were required to get there
```
Comment (MohammadMohsinKhan) [inline docs/VnVPlan/VnVPlan.tex:657]: developer or user?

File: docs/VnVPlan/VnVPlan.tex [hunk 5]
```diff
@@ -549,8 +549,317 @@ \subsection{Tests for Nonfunctional Requirements}
...
+  Initial State: Application on home page, ready for use
+            
+  Input: Tester engagement
+            
+  Output: All identified instances of foreign languages used
+            
+  How test will be performed: 
+  \begin{enumerate}
+    \item The tester will look through each page, and test functionality
+    \item They will verify that all pages and prompts are in the correct language
```
Comment (MohammadMohsinKhan) [inline docs/VnVPlan/VnVPlan.tex:675]: maybe mention the correct language is English (US)

File: docs/VnVPlan/VnVPlan.tex (hunk 1 above)
Comment (DFong298) [reply docs/VnVPlan/VnVPlan.tex:561]: No, just need that _a_ developer is in charge of it

**Review by lucasichen — CHANGES_REQUESTED** [review]

File: docs/VnVPlan/VnVPlan.tex [hunk 6]
```diff
@@ -549,8 +549,318 @@ \subsection{Tests for Nonfunctional Requirements}
...
+Initial State: Application on home page, ready for use
+					
+Input/Condition: Tester engagement
+					
+Output/Result: List of all identified inconsistencies that are found that do not adhere to SRS defined requirements.
+					
+How test will be performed: 
+\begin{enumerate}
+  \item A developer will review the code to make sure the look and feels requirements were implemented
+  \item They will then look at the application as well to ensure that the implementation was made correctly
```
Comment (lilweege) [inline docs/VnVPlan/VnVPlan.tex:584]:
```suggestion
  \item They will then check that the application to ensure that it adheres to all look and feel requirements
```

File: docs/VnVPlan/VnVPlan.tex [hunk 7]
```diff
@@ -549,8 +549,318 @@ \subsection{Tests for Nonfunctional Requirements}
...
+
+NFR: LF-AR1, LF-AR2, LF-AR3, LF-SR1, LF-SR2, LF-SR3, LF-SR4
+
+Type: Manual, Front End Review
+					
+Initial State: Application on home page, ready for use
+					
+Input/Condition: Tester engagement
+					
+Output/Result: List of all identified inconsistencies that are found that do not adhere to SRS defined requirements.
```
Comment (lilweege) [inline docs/VnVPlan/VnVPlan.tex:579]: Can we list them explicitly? I.e. For each NFR we should mention what we should be adhering to.

File: docs/VnVPlan/VnVPlan.tex [hunk 8]
```diff
@@ -549,8 +549,318 @@ \subsection{Tests for Nonfunctional Requirements}
...
+Initial State: Application on home page, ready for use
+					
+Input: Tester engagement
+					
+Output: List of missing accessibility considerations that have been overlooked
+					
+How test will be performed: 
+\begin{enumerate}
+  \item The developer will review the code and application to ensure that accessibility considerations have been implemented
+  \item They will verify that all images have alt text
```
Comment (lilweege) [inline docs/VnVPlan/VnVPlan.tex:602]:
```suggestion
  \item They will verify that all images have alt text
  \item They will verify that all text has sufficient contrast and sizing using Google Chrome devtools
```

File: docs/VnVPlan/VnVPlan.tex [hunk 9]
```diff
@@ -549,8 +549,318 @@ \subsection{Tests for Nonfunctional Requirements}
...
+            
+  Output/Result: Subjective feedback for how well application scores on different requirements relating to user experience
+            
+  How test will be performed: 
+  \begin{enumerate}
+    \item A user will use the application, going through the onboarding process
+    and doing a demo of the application
+    \item After a full demo, the user will be asked to provide feedback on
+    different aspects of usability, such as ease of use, app clarity, and prompt
+    politeness.
```
Comment (lilweege) [inline docs/VnVPlan/VnVPlan.tex:640]: Can we specify what they're being asked? We should put the exact wording of the question such as "On a scale of 1 to 10, how would you rate X". Then the output/result section can be "User feedback scores in categories X, Y, Z on a scale of 1 to 10"

Comment (lilweege) [reply docs/VnVPlan/VnVPlan.tex:640]: Include one question for each requirement that this test is for

File: docs/VnVPlan/VnVPlan.tex [hunk 10]
```diff
@@ -549,8 +549,318 @@ \subsection{Tests for Nonfunctional Requirements}
...
+            
+  Initial State: Application on home page, ready for use
+            
+  Input: Tester engagement
+            
+  Output: A count for the number of clicks required to navigate to different functionality
+            
+  How test will be performed: 
+  \begin{enumerate}
+    \item A list of commonly used functions will be created (System is not large, therefore this can be done manually)
```
Comment (lilweege) [inline docs/VnVPlan/VnVPlan.tex:657]: What are the commonly used functions? We need to list them out.
Here's what I had in mind (only really these two matter):
- Make a submission
- View a generated report

File: docs/VnVPlan/VnVPlan.tex [hunk 11]
```diff
@@ -549,8 +549,318 @@ \subsection{Tests for Nonfunctional Requirements}
...
+            
+  Initial State: Application on home page, ready for use
+            
+  Input: Tester engagement
+            
+  Output: All identified instances of foreign languages used
+            
+  How test will be performed: 
+  \begin{enumerate}
+    \item The tester will look through each page, and test functionality
```
Comment (lilweege) [inline docs/VnVPlan/VnVPlan.tex:675]: They don't need to test functionality, just check that it's in English

File: docs/VnVPlan/VnVPlan.tex [hunk 12]
```diff
@@ -549,8 +549,317 @@ \subsection{Tests for Nonfunctional Requirements}
...
+\begin{description}
+  \item[Front End Component Review:] We will engage a developer to look through
+  the code and interface and ensure components are implemented as they are
+  defined in the SRS.
+  \item[Accessibility Check:] A developer will review if accessibility tools
+  exist or not within the application. Important tools are screen readers and
+  alt text for images. Furthermore, the developer must check for sufficient
+  colour contrast between components.
+\end{description}
+
```
[the remark this answers is not in the document]
Comment (lucasichen) [reply docs/VnVPlan/VnVPlan.tex:565]: fixed

**Review by lilweege — COMMENTED** [review]
More review to come, only really looked at the first couple so far

File: docs/VnVPlan/VnVPlan.tex [hunk 13]
```diff
@@ -549,15 +549,330 @@ \subsection{Tests for Nonfunctional Requirements}
...
+            
+  Initial State: N/A
+            
+  Input/Condition: Tester engagement
+            
+  Output/Result: The way to find old iterations/versions of project
+            
+  How test will be performed: 
+  \begin{enumerate}
+    \item A tester must check if old iterations are stored properly (on Github, for example)
```
Comment (lilweege) [inline docs/VnVPlan/VnVPlan.tex:702]: Old iteration of what? We should store old models with appropriate semantic versioning.

File: docs/VnVPlan/VnVPlan.tex [hunk 14]
```diff
@@ -549,15 +549,330 @@ \subsection{Tests for Nonfunctional Requirements}
...
+  Initial State: Application ready for use
+
+  Input/Condition: Tester engagement
+
+  Output/Result: List of all identified components that have data stored 
+
+  How test will be performed:
+  \begin{enumerate}
+    \item A tester will enable logging in the components of the application
+    \item They will do a run through of the different functionalities
```
Comment (lilweege) [inline docs/VnVPlan/VnVPlan.tex:745]:
```suggestion
    \item They will do a run through of the different functionalities, including making a submission and analyzing a generated report file
```

File: docs/VnVPlan/VnVPlan.tex [hunk 15]
```diff
@@ -549,15 +549,330 @@ \subsection{Tests for Nonfunctional Requirements}
...
+    \item With each function, the tester will check logs and file systems and
```
Comment (lilweege) [inline docs/VnVPlan/VnVPlan.tex:746]:
```suggestion
    \item With each function, the tester will check logs and filesystem, and then
```

File: docs/VnVPlan/VnVPlan.tex [hunk 16]
```diff
@@ -549,15 +549,330 @@ \subsection{Tests for Nonfunctional Requirements}
...
+  
+  NFR: SR-P1
+
+  Type: Manual, Data Retention Audit
+
+  Initial State: Application ready for use
+
+  Input/Condition: Tester engagement
+
+  Output/Result: List of all identified components that have data stored
```
Comment (lilweege) [inline docs/VnVPlan/VnVPlan.tex:740]:
```suggestion
  Output/Result: A list of all identified user-generated or personally identifiable data
```

File: docs/VnVPlan/VnVPlan.tex [hunk 17]
```diff
@@ -549,15 +549,330 @@ \subsection{Tests for Nonfunctional Requirements}
...
+  Input/Condition: Tester engagement
+
+  Output/Result: List of all identified components that have data stored 
+
+  How test will be performed:
+  \begin{enumerate}
+    \item A tester will enable logging in the components of the application
+    \item They will do a run through of the different functionalities
+    \item With each function, the tester will check logs and file systems and
+    search for any type of data being stored
```
Comment (lilweege) [inline docs/VnVPlan/VnVPlan.tex:747]:
```suggestion
    search for any user-generated or personally identifiable data
```

File: docs/VnVPlan/VnVPlan.tex [hunk 18]
```diff
@@ -549,15 +549,330 @@ \subsection{Tests for Nonfunctional Requirements}
...
+  
+  NFR: SR-P2 
+
+  Type: Manual 
+
+  Initial State: Application ready for use
+
+  Input/Condition: Backend endpoint invocation
+
+  Output/Result: Identified components that do not adhere to security requirements
```
Comment (lilweege) [inline docs/VnVPlan/VnVPlan.tex:761]:
```suggestion
  Output/Result: A list of exposed endpoints that do not support in-transit encryption
```

File: docs/VnVPlan/VnVPlan.tex [hunk 19]
```diff
@@ -549,15 +549,330 @@ \subsection{Tests for Nonfunctional Requirements}
...
+  Initial State: Application ready for use
+
+  Input/Condition: Tester engagement
+  
+  Output/Result: List of vulnerabilites and breach of regulations
+
+  How test will be performed:
+  \begin{enumerate}
+    \item A tester will inspect the code repository and system application against laws and regulations
+    \item All discrepancies between the law and the system will be identified, and updated promptly.
```
Comment (lilweege) [inline docs/VnVPlan/VnVPlan.tex:785]: This isn't part of the test
```suggestion
```

File: docs/VnVPlan/VnVPlan.tex [hunk 20]
```diff
@@ -549,15 +549,330 @@ \subsection{Tests for Nonfunctional Requirements}
...
+
+  Initial State: Application ready for use
+
+  Input/Condition: Tester engagement
+  
+  Output/Result: List of vulnerabilites and breach of regulations
+
+  How test will be performed:
+  \begin{enumerate}
+    \item A tester will inspect the code repository and system application against laws and regulations
```
Comment (lilweege) [inline docs/VnVPlan/VnVPlan.tex:784]:
```suggestion
    \item A third party tester will inspect the code repository and system application against laws and regulations, noting any component of the system that may potentially infringe upon any data or privacy laws, intellectual property, or academic integrity
```

File: docs/VnVPlan/VnVPlan.tex [hunk 21]
```diff
@@ -549,15 +549,330 @@ \subsection{Tests for Nonfunctional Requirements}
...
+  
+  NFR: CR-L1, CR-L2, CR-L3, CR-S2, CR-S5
+
+  Type: Manual, Legal Audit
+
+  Initial State: Application ready for use
+
+  Input/Condition: Tester engagement
+  
+  Output/Result: List of vulnerabilites and breach of regulations
```
Comment (lilweege) [inline docs/VnVPlan/VnVPlan.tex:780]:
```suggestion
  Output/Result: List of components of the system that breach any regulations, with an accompanying explanation of the how the breach could manifest itself
```

File: docs/VnVPlan/VnVPlan.tex [hunk 22]
```diff
@@ -549,15 +549,330 @@ \subsection{Tests for Nonfunctional Requirements}
...
+
+  Type: Manual, Code Inspection
+
+  Initial State: Development temporarily paused
+
+  Output/Result: Indication of components not adhering to software development standards
+
+  How test will be performed:
+  \begin{enumerate}
+    \item Team members will regularly inspect the code and ensure all components follow software development standards
```
Comment (lilweege) [inline docs/VnVPlan/VnVPlan.tex:835]:
```suggestion
    \item Team members will regularly inspect the code and ensure all components follow software development standards, including SOLID principles and the existence of documentation and tests
```

File: docs/VnVPlan/VnVPlan.tex [hunk 23]
```diff
@@ -549,15 +549,330 @@ \subsection{Tests for Nonfunctional Requirements}
...
+  \item[Documentation Review:] A tester will look through the webpage, and
+  ensure that documentation exists. Then, they must check to make sure the
+  documentation is accurate, and not misleading. 
+\end{description}
+
+\paragraph{Validate Program, help, and training documentation exists and is accurate}
+\begin{enumerate}
+  \item{test-D-1}
+  
+  NFR: MS-S1, UH-L2, OE-P2
```
Comment (lilweege) [inline docs/VnVPlan/VnVPlan.tex:853]: I think MS-S1 should be included in the previous test, it is more so about internal API documentation than user documentation.
```suggestion
  NFR: UH-L2, OE-P2
```

File: docs/VnVPlan/VnVPlan.tex [hunk 24]
```diff
@@ -549,15 +549,330 @@ \subsection{Tests for Nonfunctional Requirements}
...
+    \item A tester will look at the 'issues' section of the github repository 
+    \item They will check that there are bugs that exist as issues
+    \item If there aren't, they will consult the team and ask if developers are
+    issuing bugs or if there are simply no bugs
+  \end{enumerate}
+
+
+  \item{test-DS-2}
+  
+  NFR: CR-SC1
```
Comment (lilweege) [inline docs/VnVPlan/VnVPlan.tex:825]:
```suggestion
  NFR: MS-S1, CR-SC1
```

File: docs/VnVPlan/VnVPlan.tex (hunk 23 above)
Comment (lilweege) [reply docs/VnVPlan/VnVPlan.tex:853]: @MohammadMohsinKhan FYI

File: docs/VnVPlan/VnVPlan.tex [hunk 25]
```diff
@@ -549,15 +549,330 @@ \subsection{Tests for Nonfunctional Requirements}
...
+
+  Initial State: Application ready for use
+
+  Output/Result: Certification that documentation is valid
+
+  How test will be performed:
+  \begin{enumerate}
+    \item A developer will review the documentation on the application. They must first ensure that documentation exists, and is easy to find
+    \item They must then verify that the documentation is correct, and complete
+    \item Anything missing or incorrect about the documentation will be brought up to the team immediately.
```
Comment (lilweege) [inline docs/VnVPlan/VnVPlan.tex:865]:
```suggestion
    \item A developer will navigate to the user documentation and help pages within the application. They must first ensure that documentation exists, and is easy to find
    \item They must then verify that the documentation is correct, complete, uses clear non-technical language, and is helpful for users
```

**Review by lilweege — CHANGES_REQUESTED** [review]

File: docs/VnVPlan/VnVPlan.tex (hunk 7 above)
Comment (lilweege) [reply docs/VnVPlan/VnVPlan.tex:579]: Addressed in 91e5c568de4fc8efc5f61fffd99b982b9db10ebc

File: docs/VnVPlan/VnVPlan.tex [hunk 26]
```diff
@@ -549,15 +549,337 @@ \subsection{Tests for Nonfunctional Requirements}
...
+\begin{enumerate}
+  \item{test-V-1}
+
+  NFR: MS-M1, OR-R2
+  
+  Type: Manual
+            
+  Initial State: N/A
+            
+  Input/Condition: Tester engagement
```
Comment (lilweege) [inline docs/VnVPlan/VnVPlan.tex:705]: What does this mean lol

File: docs/VnVPlan/VnVPlan.tex (hunk 11 above)
Comment (DFong298) [reply docs/VnVPlan/VnVPlan.tex:684]: How will they seek every prompt without going through functionality?

File: docs/VnVPlan/VnVPlan.tex (hunk 10 above)
Comment (DFong298) [reply docs/VnVPlan/VnVPlan.tex:666]: I literally said a list of commonly used functions will be created

File: docs/VnVPlan/VnVPlan.tex (hunk 9 above)
Comment (DFong298) [reply docs/VnVPlan/VnVPlan.tex:649]: This belongs in a survey we have yet to make

File: docs/VnVPlan/VnVPlan.tex (hunk 26 above)
Comment (DFong298) [reply docs/VnVPlan/VnVPlan.tex:705]: The tester will be doing things with the app to test it manually

File: docs/VnVPlan/VnVPlan.tex (hunk 10 above)
Comment (lilweege) [reply docs/VnVPlan/VnVPlan.tex:666]: You can't just say the list will be created, we have to actually have the list. That's the point of the test.

Comment (lilweege) [reply docs/VnVPlan/VnVPlan.tex]: Addressed in 917cc17c45140145e766d64cb16526099f24393d

Comment (DFong298) [reply docs/VnVPlan/VnVPlan.tex]: As in, the list will be made when we do the actual tests

File: docs/VnVPlan/VnVPlan.tex (hunk 11 above)
Comment (lilweege) [reply docs/VnVPlan/VnVPlan.tex:688]: They don't have to necessarily perform the function, just go to the page

Comment (lilweege) [reply docs/VnVPlan/VnVPlan.tex]: Addressed in 5f4b6915d17ae7f03a474df7db5470520e5bac19

File: docs/VnVPlan/VnVPlan.tex (hunk 9 above)
Comment (lilweege) [reply docs/VnVPlan/VnVPlan.tex:649]: It needs to be *specific*. This document is meant to include that survey if there is one

Comment (lilweege) [reply docs/VnVPlan/VnVPlan.tex:649]: {39F50886-B2B5-4A13-963B-719318B4FA75}

**Review by lilweege — DISMISSED** [review]

**Review by lucasichen — APPROVED** [review]

**Review by MohammadMohsinKhan — APPROVED** [review]

**Pull request merged**